File: pe-core/src/main/java/org/saar/core/common/r2d/MeshWriter2D.java
package org.saar.core.common.r2d;

import org.saar.core.mesh.writers.ElementsMeshWriter;

public class MeshWriter2D implements ElementsMeshWriter<Vertex2D> {

    private final MeshPrototype2D prototype;

    public MeshWriter2D(MeshPrototype2D prototype) {
        this.prototype = prototype;
    }

    @Override
    public void writeVertex(Vertex2D vertex) {
        this.prototype.getPositionBuffer().getWriter().write2f(vertex.getPosition2f());
        this.prototype.getColourBuffer().getWriter().write3f(vertex.getColour3f());
    }

    @Override
    public void writeIndex(int index) {
        this.prototype.getIndexBuffer().getWriter().writeInt(index);
    }
}

File: pe-core/src/main/java/org/saar/core/common/r2d/MeshWriter2D.kt
package org.saar.core.common.r2d

import org.saar.core.mesh.builder.IndexedMeshWriter
import org.saar.core.mesh.builder.VertexMeshWriter
import org.saar.lwjgl.util.DataWriter

class MeshWriter2D(
    private val positionWriter: DataWriter,
    private val colourWriter: DataWriter,
    private val indexWriter: DataWriter,
) : VertexMeshWriter<Vertex2D>, IndexedMeshWriter {

    override fun writeVertex(vertex: Vertex2D) {
        this.positionWriter.write2f(vertex.position2f)
        this.colourWriter.write3f(vertex.colour3f)
    }

    override fun writeIndex(index: Int) = this.indexWriter.writeInt(index)
}